GPII-4102: Block egress traffic

stepanstipl opened the pull request:
This PR adds network policies to block egress external traffic unless it's going via the egress gateway, it's part of work on controlling egress traffic - [GPII-4102](https://issues.gpii.net/browse/GPII-4102). This will effectively bring us back to the state before GKE Istio 1.1, when any egress traffic was denied by default.

**PR Changes:**
- Block egress traffic from `gpii` namespace - egress traffic is blocked by default, there are several exceptions:
  - Egress to Istio control plane is allowed - this is required for traffic routed via Egress gateway and other Istio traffic to istio control plane
  - Egress traffic within `gpii` namespace is allowed (traffic is already controlled by ingress rules on corresponding services)
  - Egress traffic to Kube DNS service is allowed
  - Egress traffic to GCP metadata is allowed on port 8181

**Deployment:**
This is a non-disruptive deploy. (All the traffic is already routed via Egress gateway).

**Testing:**
- I've confirmed traffic is truly routed via Egress gateway and all other traffic blocked after applying blocking Network Policies.

natarajaya commented on gcp/modules/gpii-istio/network_policy.tf:
```diff
@@ -5,9 +5,57 @@ resource "kubernetes_network_policy" "deny-default" {
+      # Allow egress traffic Compute Metadata
+      to {
+        ip_block {
+          cidr = "10.16.0.0/20"
```
There are more networks in [cluster_subnets](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gke-network/main.tf#L22), do they need to be included here as well?

stepanstipl replied:
Nope, only the IPs of the nodes, because that's what the metadata traffic is being natted to (the other IPs are for services and pods).

natarajaya replied:
Ok, makes sense. Had to look into exekube's `gke-network` to understand what these extra networks are.

natarajaya commented on gcp/modules/gpii-istio/network_policy.tf:
```diff
@@ -5,9 +5,57 @@ resource "kubernetes_network_policy" "deny-default" {
+        protocol = "TCP"
+      }
+    }
 
```
What about traffic to GCP APIs (metrics, trace)? Also traffic to flowmanager public endpoint needs to be allowed for SSL check to work, in case that PR merges first.

stepanstipl commented:
Thanks for reviews @natarajaya, I believed I addressed all the points. I'll add some details about the traffic routing directly to the https://github.com/gpii-ops/gpii-infra/pull/548 PR.

natarajaya commented:
LGTM
